source/include/core_sntp_client.h

```diff
@@ -0,0 +1,463 @@
+ * @p clockOffSetSec, to apply correction to the system clock with a "slew rate"
+ * (that is higher than the SNTP polling rate).
+ * If the application can accept sudden jump in time (forward or backward), then
+ * the "step" discipline methodology can be used to directly update the system
+ * clock with the current server time, @p pServerTime, every time the coreSNTP
+ * library calls the interface.
+ *
+ * @return `true` if setting system time is successful; otherwise `false` to
+ * represent failure.
+ */
+typedef bool ( * SntpSetTime_t )( const char * pTimeServer,
+                                  const SntpTimestamp_t * pServerTime,
+                                  int32_t clockOffsetSec );
+
+/**
+ * @brief The default UDP port supported by SNTP/NTP servers for client-server
+ * communication.
+ *
+ * @note It is possible for a server to use a different port number than
+ * the default port when using the Network Time Security protocol as the security
+ * mechanism for SNTP communication. For more information, refer to Section 4.1.8
+ * of [RFC 8915](https://tools.ietf.org/html/rfc8915).
+ */
+#define SNTP_DEFAULT_SERVER_PORT    ( 123U )
+
+/**
+ * @brief core_sntp_struct_types
+ * @brief Structure representing information for a time server.
+ */
+typedef struct SntpServerInfo
+{
+    const char * pServerName; /**<@brief The time server endpoint. */
+    uint16_t port;            /**<@brief The UDP port supported by the server
+                               * for SNTP/NTP communication. */
+} SntpServerInfo_t;
+
+/**
+ * @ingroup core_sntp_struct_types
+ * @typedef NetworkContext_t
+ * @brief A user-defined type for context that is passed to the transport interface functions.
+ * It MUST be defined by the user to use the library.
+ * It is of incomplete type to allow user to defined to the the needs of their transport
+ * interface implementation.
+ */
+struct NetworkContext;
+typedef struct NetworkContext NetworkContext_t;
+
+/**
+ * @ingroup core_sntp_callback_types
+ * @brief Interface for user-defined function to send data to the network
+ * over User Datagram Protocol (UDP).
+ *
+ * @param[in,out] pNetworkContext The user defined NetworkContext_t which
+ * is opaque to the coreSNTP library.
+ * @param[in] pTimeServer The time server to send the data to.
+ * @param[in] pBuffer The buffer containing the data to send over the network.
+ * @param[in] bytesToSend The size of data in @p pBuffer to send.
+ *
+ * @return The function SHOULD return one of the following integer codes:
+ * 0 when all data is successfully transmitted over the network.
+ * -1 when only partial data is sent over the network, and operation can be
+ * re-tried to send the remaining data.
+ * -2 when the send operation failed to send any data, and operation
+ * cannot be re-tried.
+ */
+typedef int32_t ( * UdpTransportSendTo_t )( NetworkContext_t * pNetworkContext,
+                                            const SntpServerInfo_t * pTimeServer,
```

Comment:
I'm still going to say that `pTimeServer` can be embedded into `pNetworkContext`, but this can be done later on if we fail to see a reason as to why this extra transport interface for UDP is necessary.

Reply:
The library is responsible for selecting the time server for requesting time from the list of time servers provided by the application. Therefore, passing the time server to the application-defined callback is necessary here